File: code/futureforge/group-34/tries.py


# A trie (or prefix tree) is a tree-like data structure used for efficiently storing and retrieving a dynamic set of strings. 
# Each node in a trie represents a single character of a string, and paths from the root to a node represent prefixes of strings.

class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """Inserts a word into the trie."""
        current_node = self.root
        for char in word:
            if char not in current_node.children:
                current_node.children[char] = TrieNode()
            current_node = current_node.children[char]
        current_node.is_end_of_word = True

    def search(self, word):
        """Checks if a word exists in the trie."""
        current_node = self.root
        for char in word:
            if char not in current_node.children:
                return False
            current_node = current_node.children[char]
        return current_node.is_end_of_word

    def starts_with(self, prefix):
        """Checks if there is any word in the trie that starts with the given prefix."""
        current_node = self.root
        for char in prefix:
            if char not in current_node.children:
                return False
            current_node = current_node.children[char]
        return True

# Example usage of Trie
trie = Trie()

trie.insert("apple")
trie.insert("apricot")
trie.insert("banana")

print(trie.search("apple"))  # Output: True
print(trie.search("apricot"))  # Output: True
print(trie.search("orange"))  # Output: False

print(trie.starts_with("app"))  # Output: True
print(trie.starts_with("ora"))  # Output: False

File: code/futureforge/group-35/tries.js
//CEO's Work 20/CSC/119

// A trie (or prefix tree) is a tree-like data structure used for efficiently storing and retrieving a dynamic set of strings. 
// Each node in a trie represents a single character of a string. The structure of a trie allows for quick lookups and prefix-based searches.


class TrieNode {
    constructor() {
        this.children = {};  // Map of child nodes (character to TrieNode)
        this.isEndOfWord = false;  // Flag to indicate end of a word
    }
}

class Trie {
    constructor() {
        this.root = new TrieNode();  // Root node of the trie
    }

    insert(word) {
        let current = this.root;

        // Traverse through each character in the word
        for (let char of word) {
            if (!current.children[char]) {
                // Create a new node if the character doesn't exist
                current.children[char] = new TrieNode();
            }
            // Move to the child node corresponding to the character
            current = current.children[char];
        }

        // Mark the end of the inserted word
        current.isEndOfWord = true;
    }

    search(word) {
        let current = this.root;

        // Traverse through each character in the word
        for (let char of word) {
            if (!current.children[char]) {
                // If a character is not found, the word does not exist in the trie
                return false;
            }
            // Move to the child node corresponding to the character
            current = current.children[char];
        }

        // Check if we reached the end of a valid word
        return current.isEndOfWord;
    }

    startsWith(prefix) {
        let current = this.root;

        // Traverse through each character in the prefix
        for (let char of prefix) {
            if (!current.children[char]) {
                // If a character is not found, the prefix does not exist in the trie
                return false;
            }
            // Move to the child node corresponding to the character
            current = current.children[char];
        }

        // Prefix found in the trie
        return true;
    }
}

// Example usage of Trie
const trie = new Trie();

trie.insert("apple");
trie.insert("apricot");
trie.insert("banana");
trie.insert("orange");

console.log(trie.search("apple"));  // Output: true
console.log(trie.search("apricot"));  // Output: true
console.log(trie.search("banana"));  // Output: true
console.log(trie.search("orange"));  // Output: true
console.log(trie.search("grape"));  // Output: false

console.log(trie.startsWith("ap"));  // Output: true
console.log(trie.startsWith("ban"));  // Output: true
console.log(trie.startsWith("ora"));  // Output: true
console.log(trie.startsWith("pea"));  // Output: false